IndexOutOfBoundsException in getElementsByTail

Executing the validator with the included resource and the following params results in an lang.IndexOutOfBoundsException
```bash
java -jar validator.jar -ig
ca.on.oh-erec-ig#0.12.4-alpha1.0.0
./Communication_Doc-Reference-2_public.json
```

[Communication_Doc-Reference-2_public.json](https://github.com/user-attachments/files/22119602/Communication_Doc-Reference-2_public.json)

```
Exception in thread "main" org.hl7.fhir.exceptions.FHIRException: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
    at org.hl7.fhir.validation.ValidationEngine.validate(ValidationEngine.java:693)
    at org.hl7.fhir.validation.service.ValidationService.validateSources(ValidationService.java:264)
    at org.hl7.fhir.validation.cli.tasks.ValidateTask.executeTask(ValidateTask.java:68)
    at org.hl7.fhir.validation.cli.ValidatorCli.readParamsAndExecuteTask(ValidatorCli.java:375)
    at org.hl7.fhir.validation.cli.ValidatorCli.readGlobalParamsAndExecuteTask(ValidatorCli.java:192)
    at org.hl7.fhir.validation.cli.ValidatorCli.main(ValidatorCli.java:254)
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
    at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
    at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
    at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
    at java.base/java.util.Objects.checkIndex(Objects.java:361)
    at java.base/java.util.ArrayList.get(ArrayList.java:427)
    at org.hl7.fhir.validation.instance.InstanceValidator.getElementByTail(InstanceValidator.java:7951)
    at org.hl7.fhir.validation.instance.InstanceValidator.checkChildByDefinition(InstanceValidator.java:7373)
    at org.hl7.fhir.validation.instance.InstanceValidator.checkChild(InstanceValidator.java:7091)
```

--- a/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
+++ b/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
@@ -7370,11 +7370,17 @@ public class InstanceValidator extends BaseValidator implements IResourceValidat
         p = this.context.fetchResource(StructureDefinition.class, typeProfile);
         if (rule(errors, NO_RULE_DATE, IssueType.STRUCTURE, ei.line(), ei.col(), ei.getPath(), p != null, I18nConstants.VALIDATION_VAL_UNKNOWN_PROFILE, typeProfile)) {
           List<ValidationMessage> profileErrors = new ArrayList<ValidationMessage>();
-          validateElement(valContext, profileErrors, p, getElementByTail(p, tail), profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
-          if (hasErrors(profileErrors))
-            badProfiles.put(typeProfile, profileErrors);
-          else
-            goodProfiles.put(typeProfile, profileErrors);
+          try {
+            ElementDefinition tailElement = getElementByTail(p, tail);
+            validateElement(valContext, profileErrors, p, tailElement, profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
+            if (hasErrors(profileErrors))
+              badProfiles.put(typeProfile, profileErrors);
+            else
+              goodProfiles.put(typeProfile, profileErrors);
+          } catch (DefinitionException exception) {
+            log.error("Error getting tail element", exception);
+            ok = false;
+          }
         } else {
           ok = false;
         }
@@ -7947,7 +7953,7 @@ public class InstanceValidator extends BaseValidator implements IResourceValidat
   }
 
   private ElementDefinition getElementByTail(StructureDefinition p, String tail) throws DefinitionException {
-    if (tail == null)
+    if (tail == null && !p.getSnapshot().getElement().isEmpty())
       return p.getSnapshot().getElement().get(0);
     for (ElementDefinition t : p.getSnapshot().getElement()) {
       if (tail.equals(t.getId()))

--- a/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
+++ b/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
@@ -7368,17 +7368,14 @@ public class InstanceValidator extends BaseValidator implements IResourceValidat
           url = url.substring(0, url.indexOf("#"));
         }
         p = this.context.fetchResource(StructureDefinition.class, typeProfile);
-        if (rule(errors, NO_RULE_DATE, IssueType.STRUCTURE, ei.line(), ei.col(), ei.getPath(), p != null, I18nConstants.VALIDATION_VAL_UNKNOWN_PROFILE, typeProfile)) {
-          if (!p.getSnapshot().hasElement()) {
-            rule(errors, null, IssueType.INVALID, stack, false, I18nConstants.SNAPSHOT_IS_EMPTY, p.getVersionedUrl());
-          } else {
-            List<ValidationMessage> profileErrors = new ArrayList<ValidationMessage>();
-            validateElement(valContext, profileErrors, p, getElementByTail(p, tail), profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
-            if (hasErrors(profileErrors))
-              badProfiles.put(typeProfile, profileErrors);
-            else
-              goodProfiles.put(typeProfile, profileErrors);
-          }
+        if (rule(errors, NO_RULE_DATE, IssueType.STRUCTURE, ei.line(), ei.col(), ei.getPath(), p != null, I18nConstants.VALIDATION_VAL_UNKNOWN_PROFILE, typeProfile)
+          && rule(errors, NO_RULE_DATE, IssueType.STRUCTURE, ei.line(), ei.col(), ei.getPath(), p.getSnapshot().hasElement(), I18nConstants.VALIDATION_VAL_PROFILE_NOSNAPSHOT, typeProfile)) {
+          List<ValidationMessage> profileErrors = new ArrayList<ValidationMessage>();
+          validateElement(valContext, profileErrors, p, getElementByTail(p, tail), profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
+          if (hasErrors(profileErrors))
+            badProfiles.put(typeProfile, profileErrors);
+          else
+            goodProfiles.put(typeProfile, profileErrors);
         } else {
           ok = false;
         }

--- a/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
+++ b/org.hl7.fhir.validation/src/main/java/org/hl7/fhir/validation/instance/InstanceValidator.java
@@ -7369,17 +7369,15 @@ public class InstanceValidator extends BaseValidator implements IResourceValidat
         }
         p = this.context.fetchResource(StructureDefinition.class, typeProfile);
         if (rule(errors, NO_RULE_DATE, IssueType.STRUCTURE, ei.line(), ei.col(), ei.getPath(), p != null, I18nConstants.VALIDATION_VAL_UNKNOWN_PROFILE, typeProfile)) {
-          List<ValidationMessage> profileErrors = new ArrayList<ValidationMessage>();
-          try {
-            ElementDefinition tailElement = getElementByTail(p, tail);
-            validateElement(valContext, profileErrors, p, tailElement, profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
+          if (!p.getSnapshot().hasElement()) {
+            rule(errors, null, IssueType.INVALID, stack, false, I18nConstants.SNAPSHOT_IS_EMPTY, p.getVersionedUrl());
+          } else {
+            List<ValidationMessage> profileErrors = new ArrayList<ValidationMessage>();
+            validateElement(valContext, profileErrors, p, getElementByTail(p, tail), profile, checkDefn, resource, ei.getElement(), type, localStack, thisIsCodeableConcept, checkDisplay, thisExtension, pct, mode); // we don't track ok here
             if (hasErrors(profileErrors))
               badProfiles.put(typeProfile, profileErrors);
             else
               goodProfiles.put(typeProfile, profileErrors);
-          } catch (DefinitionException exception) {
-            log.error("Error getting tail element", exception);
-            ok = false;
           }
         } else {
           ok = false;

